### DISCUSSION TITLE
docs routing with less fragile URLs?

### DISCUSSION THREAD

**Discussion by hatton** [facebook/docusaurus#8180] [Q&A] [answered]
We publish desktop software (so various versions out there at any one time), and want to point to Docusaurus pages from the UI. However, the way we've done the URLs (auto sidebar+folders) is fragile: 

`https://docs.bloomlibrary.org/Advanced-Topics-for-Large-Book-Projects/Technical-Notes/ePUB-notes`

This is fragile because if we move that article to somewhere else in the structure, the link will now fail for users of older versions of the application. In our context, we need to avoid creating and maintaining server redirects as that would be just another point of fragility.

I see two approaches but don't know how to do either:

1) Switch to a slug-based URL scheme: `https://docs.bloomlibrary.org/ePUB-notes`

2) Have some kind of permalink alternative. I.e., our application's code would have non-changing `https://docs.bloomlibrary.org/12345` but that takes you to the current full URL `https://docs.bloomlibrary.org/Advanced-Topics-for-Large-Book-Projects/Technical-Notes/ePUB-notes`?

I fear this is covered in [routing](https://docusaurus.io/docs/advanced/routing) or [docs plugin](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs) but I'm just missing it. 

Thanks for any help! ([repo here](https://github.com/bloomBooks/bloom-docs))

**Comment by slorber** [comment, marked as answer]
There's a `slug: /ePUB-notes` frontmatter that you can use with any of our content plugins.

https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs#slug

**Reply by hatton** [reply]
Right, I do have slugs, but somehow I'm not configured to get any benefit from them.  E.g.

`https://docs.bloomlibrary.org/Advanced-Topics-for-Large-Book-Projects/Technical-Notes/ePUB-notes` works, but
`https://docs.bloomlibrary.org/ePUB-notes` 404's, even though I have:

```
---
title: Notes on Publish to ePUB
sidebar_position: 5
slug: ePUB-notes
---
```

[config file](https://github.com/BloomBooks/bloom-docs/blob/master/docusaurus.config.js)

If I change the slug from `ePUB-notes` to `/ePUB-notes`, then

`https://docs.bloomlibrary.org/ePUB-notes`works, but `https://docs.bloomlibrary.org/Advanced-Topics-for-Large-Book-Projects/Technical-Notes/ePUB-notes` 404's.

So yeah, I suppose this is a solution. I don't get a way to have both kinds of URLs work, but I can put `/` on any slugs that I want to be permalinks.

**Reply by hatton** [reply]
Since my markdown is generated (from notion via [docu-notion](https://github.com/sillsdev/docu-notion)) in the end I decided to auto-generate slugs if they are missing based on the Notion page id, so everything is a permalink,  and is also human-readable if the author supplies a slug.

**Reply by slorber** [reply]
`slug: ePUB-notes` is a relative url slug and will not be resolved the same as `slug: /ePUB-notes`. 

By default it will be prefixed by the "slug of the folders"

> https://docs.bloomlibrary.org/ePUB-notesworks, but https://docs.bloomlibrary.org/Advanced-Topics-for-Large-Book-Projects/Technical-Notes/ePUB-notes 404's.

This is expected, the doc can only have one canonical URL. If you want it to be available at 2 URLs you'd have to setup redirects (preferably server-side)

> Since my markdown is generated (from notion via [docu-notion](https://github.com/sillsdev/docu-notion)) in the end I decided to auto-generate slugs if they are missing based on the Notion page id, so everything is a permalink, and is also human-readable if the author supplies a slug.

Makes sense to me. I don't know Notion much but at the end of the day you'd rather have it give you a unique URL that never change in an unexpected way. Maybe you could declare the slug directly on Notion in case users want to control the URL of a page 🤷‍♂️

Cool project 👍

**Reply by hatton** [reply]
> Maybe you could declare the slug directly on Notion in case users want to control the URL of a page

Yes, exactly:

![image](https://user-images.githubusercontent.com/8448/195953235-047b6390-844b-4d8d-a655-58d8a0f6c542.png)

Produces
```
---
title: Notes on “Publish to ePUB”
sidebar_position: 5
slug: /ePUB-notes
keywords: [ePUB, publish, ebook]
---
```

For our project, we're going to ask authors to always fill in the slug, but other projects could just go with the notion id. [docu-notion](https://github.com/sillsdev/docu-notion) will auto insert the "/" as needed.
